File: packages/shared/src/dummy-invoice-data.ts
import { InvoiceStatus } from './invoice-status.js';

export const DUMMY_DATE_MIN_DAYS_OFFSET = -2;
export const DUMMY_DATE_MAX_DAYS_OFFSET = 7;

export interface DummyInvoiceFields {
  concept: string;
  amount: number;
  dueDate: Date;
}

const CONCEPT_TEMPLATES = [
  'Suscripción SaaS',
  'Soporte técnico',
  'Consultoría integración API',
  'Implementación módulo facturación',
  'Capacitación usuarios',
  'Mantenimiento mensual',
  'Licencias anuales',
  'Desarrollo personalizado reportes',
  'Migración de datos históricos',
  'Auditoría de cumplimiento',
];

const MONTHS_ES = [
  'Enero',
  'Febrero',
  'Marzo',
  'Abril',
  'Mayo',
  'Junio',
  'Julio',
  'Agosto',
  'Septiembre',
  'Octubre',
  'Noviembre',
  'Diciembre',
];

export function startOfDay(date: Date): Date {
  const result = new Date(date);
  result.setHours(0, 0, 0, 0);
  return result;
}

export function diffDaysFromToday(dueDate: Date, today: Date = new Date()): number {
  const due = startOfDay(dueDate).getTime();
  const ref = startOfDay(today).getTime();
  return Math.round((due - ref) / (1000 * 60 * 60 * 24));
}

export function randomInt(min: number, max: number): number {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

export function addDays(date: Date, days: number): Date {
  const result = new Date(date);
  result.setDate(result.getDate() + days);
  return startOfDay(result);
}

export function randomDueDateFromToday(options?: {
  minDaysAhead?: number;
  maxDaysAhead?: number;
  today?: Date;
}): Date {
  const today = startOfDay(options?.today ?? new Date());
  const minDays = options?.minDaysAhead ?? 0;
  const maxDays = options?.maxDaysAhead ?? DUMMY_DATE_MAX_DAYS_OFFSET;
  return addDays(today, randomInt(minDays, maxDays));
}

export function randomDueDateFromTwoDaysAgo(options?: {
  maxDaysAhead?: number;
  today?: Date;
}): Date {
  const today = startOfDay(options?.today ?? new Date());
  const maxDays = options?.maxDaysAhead ?? DUMMY_DATE_MAX_DAYS_OFFSET;
  return addDays(today, randomInt(DUMMY_DATE_MIN_DAYS_OFFSET, maxDays));
}

export function generateRandomConcept(): string {
  const template = CONCEPT_TEMPLATES[randomInt(0, CONCEPT_TEMPLATES.length - 1)];
  const month = MONTHS_ES[randomInt(0, MONTHS_ES.length - 1)];
  const year = new Date().getFullYear() + randomInt(0, 1);
  return `${template} — ${month} ${year}`;
}

export function generateRandomAmount(): number {
  const steps = randomInt(5, 90);
  return steps * 10_000;
}

export function generateDummyInvoiceFields(
  mode: 'seed' | 'form',
  options?: { maxDaysAhead?: number; today?: Date },
): DummyInvoiceFields {
  const dueDate =
    mode === 'seed'
      ? randomDueDateFromTwoDaysAgo({ maxDaysAhead: options?.maxDaysAhead, today: options?.today })
      : randomDueDateFromToday({ maxDaysAhead: options?.maxDaysAhead, today: options?.today });

  return {
    concept: generateRandomConcept(),
    amount: generateRandomAmount(),
    dueDate,
  };
}

export function deriveStatusFromDueDate(dueDate: Date, today?: Date): InvoiceStatus {
  const diffDays = diffDaysFromToday(dueDate, today);

  if (diffDays >= 0) {
    return InvoiceStatus.AL_DIA;
  }
  if (diffDays >= -7) {
    return InvoiceStatus.PRIMER_RECORDATORIO;
  }
  if (diffDays >= -21) {
    return InvoiceStatus.SEGUNDO_RECORDATORIO;
  }
  return InvoiceStatus.DESACTIVADO;
}

File: backend/dotnet/src/Monolegal.Shared/DummyInvoiceData.cs
namespace Monolegal.Shared;

public static class DummyInvoiceData
{
    public const int DateMinDaysOffset = -2;
    public const int DateMaxDaysOffset = 7;

    private static readonly string[] ConceptTemplates =
    [
        "Suscripción SaaS",
        "Soporte técnico",
        "Consultoría integración API",
        "Implementación módulo facturación",
        "Capacitación usuarios",
        "Mantenimiento mensual",
        "Licencias anuales",
        "Desarrollo personalizado reportes",
        "Migración de datos históricos",
        "Auditoría de cumplimiento"
    ];

    private static readonly string[] MonthsEs =
    [
        "Enero", "Febrero", "Marzo", "Abril", "Mayo", "Junio",
        "Julio", "Agosto", "Septiembre", "Octubre", "Noviembre", "Diciembre"
    ];

    public static DateTime StartOfDay(DateTime date) =>
        date.Date;

    public static int DiffDaysFromToday(DateTime dueDate, DateTime? today = null)
    {
        var due = StartOfDay(dueDate);
        var reference = StartOfDay(today ?? DateTime.Now);
        return (int)Math.Round((due - reference).TotalDays);
    }

    public static int RandomInt(int min, int max) =>
        Random.Shared.Next(min, max + 1);

    public static DateTime AddDays(DateTime date, int days) =>
        StartOfDay(date.AddDays(days));

    public static DateTime RandomDueDateFromToday(int minDaysAhead = 0, int maxDaysAhead = DateMaxDaysOffset, DateTime? today = null)
    {
        var reference = StartOfDay(today ?? DateTime.Now);
        return AddDays(reference, RandomInt(minDaysAhead, maxDaysAhead));
    }

    public static DateTime RandomDueDateFromTwoDaysAgo(int maxDaysAhead = DateMaxDaysOffset, DateTime? today = null)
    {
        var reference = StartOfDay(today ?? DateTime.Now);
        return AddDays(reference, RandomInt(DateMinDaysOffset, maxDaysAhead));
    }

    public static string GenerateRandomConcept()
    {
        var template = ConceptTemplates[RandomInt(0, ConceptTemplates.Length - 1)];
        var month = MonthsEs[RandomInt(0, MonthsEs.Length - 1)];
        var year = DateTime.Now.Year + RandomInt(0, 1);
        return $"{template} — {month} {year}";
    }

    public static decimal GenerateRandomAmount()
    {
        var steps = RandomInt(5, 90);
        return steps * 10_000m;
    }

    public record DummyInvoiceFields(string Concept, decimal Amount, DateTime DueDate);

    public static DummyInvoiceFields GenerateDummyInvoiceFields(string mode, int? maxDaysAhead = null, DateTime? today = null)
    {
        var dueDate = mode == "seed"
            ? RandomDueDateFromTwoDaysAgo(maxDaysAhead ?? DateMaxDaysOffset, today)
            : RandomDueDateFromToday(0, maxDaysAhead ?? DateMaxDaysOffset, today);

        return new DummyInvoiceFields(
            GenerateRandomConcept(),
            GenerateRandomAmount(),
            dueDate);
    }

    public static InvoiceStatus DeriveStatusFromDueDate(DateTime dueDate, DateTime? today = null)
    {
        var diffDays = DiffDaysFromToday(dueDate, today);

        if (diffDays >= 0) return InvoiceStatus.AlDia;
        if (diffDays >= -7) return InvoiceStatus.PrimerRecordatorio;
        if (diffDays >= -21) return InvoiceStatus.SegundoRecordatorio;
        return InvoiceStatus.Desactivado;
    }
}
